cmd/basic/main.go

```diff
@@ -78,152 +82,204 @@ func main() {
+	ClientFlags    uint
+	ConnCount      int
+	DelRatio       int
+	Duration       time.Duration
+	GetRatio       int
+	KeyLength      int
+	KeyPrefix      string
+	KeySpace       int
+	KeyTTL         uint
+	Pipelines      uint
+	RngSeed        int64
+	RPS            int
+	Servers        []string
+	SetRatio       int
+	Socket         string
+	StripKeyPrefix bool
+	UseZipf        bool
+	ValueSize      uint
+	WarmPercent    int
+	ZipfS          float64 // (> 1, generally 1.01-2) pulls the power curve toward 0)
+	ZipfV          float64 // v (< keySpace) puts the main part of the curve before this number
+	tachymeter     *tachymeter.Tachymeter
 }
 
-func (l *BasicLoader) Run() {
-	var runners int
-	// TODO: should be method of surfacing errors.
-	doneChan := make(chan int, 50)
+func (conf *Config) Run() (err error) {
+	g, _ := errgroup.WithContext(context.Background())
 
-	for {
-		for runners < l.desiredConnCount {
-			go l.Worker(doneChan)
-			runners++
-		}
-		res := <-doneChan
-		if res == 0 {
-			//fmt.Println("That's a bingo!")
-		}
-		runners--
-		if *cpuprofile != "" && time.Now().After(l.stopAfter) {
+	samples := conf.RPS * conf.ConnCount
+	if samples < 1000 {
+		samples = 1000
+	}
+
+	if conf.WarmPercent > 0 {
+		err = conf.WarmCache()
+		if err != nil {
 			return
 		}
 	}
+
+	threadStats := make(chan Stats, conf.ConnCount)
+	conf.tachymeter = tachymeter.New(&tachymeter.Config{Size: samples})
+	startTime := time.Now()
+
+	for worker := 0; worker < conf.ConnCount; worker++ {
+		index := worker
+		g.Go(func() error {
+			return conf.Worker(index, threadStats)
+		})
+	}
+
+	err = g.Wait()
+	endTime := time.Now()
+	if err != nil {
+		return
+	}
+
+	conf.tachymeter.SetWallTime(time.Since(startTime))
+	close(threadStats)
+	testStats := &Stats{}
+	for stats := range threadStats {
+		testStats.Add(&stats)
+	}
+
+	report := &Report{
+		StartTime: startTime,
+		EndTime:   endTime,
+		Config:    conf,
+		Metrics:   conf.tachymeter.Calc(),
+		Stats:     testStats,
+	}
+	err = report.PrettyPrint()
+
+	return
 }
 
-func (l *BasicLoader) Timer(tag string, start time.Time) {
-	duration := time.Since(start)
-	if duration > time.Millisecond * 10 {
-		fmt.Printf("%s [%d]\n", tag, int64(time.Since(start) / time.Microsecond))
+func (conf *Config) WarmCache() error {
+	mc := mct.NewClient(conf.Servers[0], conf.Socket, conf.Pipelines, conf.KeyPrefix, conf.StripKeyPrefix)
+
+	rs := pcgr.New(conf.RngSeed, 0)
+	randR := rand.New(&rs)
+	subRS := pcgr.New(1, 0)
+
+	for keyIndex := 0; keyIndex < conf.KeySpace; keyIndex++ {
+		if randR.Intn(100) < conf.WarmPercent {
+			subRS.Seed(conf.RngSeed + int64(keyIndex))
+			key := mct.RandString(&subRS, conf.KeyLength, conf.KeyPrefix)
+
+			valSeed := new(big.Int).SetBytes([]byte(key)).Int64()
+			subRS.Seed(valSeed)
+			value := mct.RandBytes(&subRS, int(conf.ValueSize))
+
+			_, err := mc.Set(key, uint32(conf.ClientFlags), uint32(conf.KeyTTL), value)
+			if err != nil {
+				fmt.Println(err)
+				return err
+			}
+		}
```

Comment:
Cache warmer iterates over every key and probabilistically sets the value based on the desired warm percentage. This is a simple approach but maybe misleading since it does not guarantee an exact percentage of keys will be set.